src/core/utils/fileutils.cpp

```diff
@@ -318,3 +321,250 @@ void FileUtils::addImageStamp( const QString &imagePath, const QString &text )
+    QDir dir = targetInfo.dir();
+    QStringList pendingSegments;
+
+    while ( !dir.exists() && dir.cdUp() )
+    {
+      pendingSegments.prepend( QFileInfo( dir.path() ).fileName() );
+    }
+
+    QString existingCanonical = QFileInfo( dir.path() ).canonicalFilePath();
+    if ( existingCanonical.isEmpty() )
+    {
+      existingCanonical = QFileInfo( dir.path() ).absoluteFilePath();
+      if ( existingCanonical.isEmpty() )
+        return false;
+    }
+
+    // Rebuild the target path from existing directories
+    QDir rebuiltDir( existingCanonical );
+    for ( const QString &segment : pendingSegments )
+    {
+      rebuiltDir.cd( segment );
+    }
+
+    targetCanonical = rebuiltDir.absoluteFilePath( targetInfo.fileName() );
+  }
+
+  // Normalize path separators
+  projectDirCanonical = QDir::fromNativeSeparators( projectDirCanonical );
+  targetCanonical = QDir::fromNativeSeparators( targetCanonical );
+
+  // Normalize paths for Windows (case-insensitive check)
+#ifdef Q_OS_WIN
+  projectDirCanonical = projectDirCanonical.toLower();
+  targetCanonical = targetCanonical.toLower();
+#endif
+
+  // Check if the target path is equal to or within the project directory
+  bool result = targetCanonical == projectDirCanonical || targetCanonical.startsWith( projectDirCanonical + "/" );
+
+  return result;
+}
+
+QByteArray FileUtils::readFileContent( const QString &filePath )
+{
+  QByteArray content;
+
+  if ( !isWithinProjectDirectory( filePath ) )
+  {
+    qWarning() << QStringLiteral( "Security warning: Attempted to read file outside project directory: %1" ).arg( filePath );
+    return QByteArray();
+  }
+
+  QFile file( filePath );
+
+  if ( file.exists() )
+  {
+    if ( file.open( QIODevice::ReadOnly ) )
+    {
+      content = file.readAll();
+      file.close();
+    }
+    else
+    {
+      qDebug() << QStringLiteral( "Failed to read file content: %1" ).arg( file.errorString() );
+    }
+  }
+  else
+  {
+    qDebug() << QStringLiteral( "File does not exist: %1" ).arg( filePath );
+  }
+
+  return content;
+}
+
+bool FileUtils::writeFileContent( const QString &filePath, const QByteArray &content )
+{
+  if ( !isWithinProjectDirectory( filePath ) )
+  {
+    qWarning() << QStringLiteral( "Security warning: Attempted to write file outside project directory: %1" ).arg( filePath );
+    return false;
+  }
+
+  QFile file( filePath );
+  QFileInfo fileInfo( filePath );
+  QDir directory = fileInfo.dir();
+
+  bool isLikelyWritable = false;
+
+#ifdef Q_OS_ANDROID
+  QString appStorage = QStandardPaths::writableLocation( QStandardPaths::AppDataLocation );
+  isLikelyWritable = filePath.startsWith( appStorage );
+#elif defined( Q_OS_IOS )
+  QString appStorage = QStandardPaths::writableLocation( QStandardPaths::AppDataLocation );
+  isLikelyWritable = filePath.startsWith( appStorage );
+#else
+  QFileInfo dirInfo( directory.absolutePath() );
+  isLikelyWritable = dirInfo.isWritable();
+#endif
```

Comment:
Do we need this? What would be the scenario(s) where we would be within a project directory but _not_ able to write?

Reply:
The idea here was to early catch situation like Permission issues where the project directory is for example on a read-only file system … or disk full errors … chmod permissions on linux etc … but I can definitly remove this check if this makes the code more clear.